waspc/waspls/test/Wasp/LSP/CompletionTest.hs

```diff
@@ -0,0 +1,145 @@
+-- 2) In one spot in the file, add a "completion marker", which is a "|" with
+-- a "^" at the same column on the line below. The line containing "^" should be
+-- blank except for whitespace before the "^" and the "^" itself.
+--
+-- The "|", "^", and the extra line are not part of the wasp source code and
+-- are not included in the returned code. The preamble comment is included in
+-- the returned code.
+--
+-- If there is more than one completion marker in the input, only the first
+-- marker is recognized, and the rest are left in the source code.
+--
+-- === __Example__
+--
+-- @
+-- //! test/completion
+-- app todoApp {
+--   |
+--   ^
+-- }
+-- @
+--
+-- This test case checks completions after the 2 spaces on the 2nd line.
+readCompletionTest :: String -> (String, LSP.Position)
+readCompletionTest source = withPreambleAssert "//! test/completion" source (unlines code, position)
+  where
+    -- Drops the marked line AND the following line (which is blank except for the ^)
+    code = before markedLineIdx (lines source) ++ [markedLine] ++ after (markedLineIdx + 1) (lines source)
+    position = LSP.Position (fromIntegral markedLineIdx) (fromIntegral markedColIdx)
+
+    markedLine = before markedColIdx rawMarkedLine ++ after markedColIdx rawMarkedLine
+    (rawMarkedLine, markedColIdx, markedLineIdx) =
+      case find isMarkedLine candidateLines of
+        Nothing -> error "readCompletionTest: no marked line"
+        Just x -> x
+
+    -- (String, column, line) triples, where String is a line such that the
+    -- following line is spaces followed by a ^ and nothing else
+    candidateLines = mapMaybe (toCandidateLine 0) linePairs
+    -- Pairs of consecutive lines in the input
+    linePairs = zip3 (lines source) (drop 1 $ lines source) [0 ..]
+
+    -- Check if the candidate line contains a '|' at the specified column
+    isMarkedLine :: (String, Int, Int) -> Bool
+    isMarkedLine (str, col, _ln) = (length str >= col) && ((str !! col) == '|')
+
+    -- Convert a pair of lines into a candidate line (this checks if the second
+    -- line is spaces followed by a ^ and nothing else)
+    toCandidateLine :: Int -> (String, String, Int) -> Maybe (String, Int, Int)
+    toCandidateLine n (a, ['^'], ln) = Just (a, n, ln)
+    toCandidateLine n (a, ' ' : bs, ln) = toCandidateLine (n + 1) (a, bs, ln)
+    toCandidateLine _ _ = Nothing
+
+    withPreambleAssert :: String -> String -> a -> a
+    withPreambleAssert preamble str x
+      | (preamble ++ "\n") `isPrefixOf` str = x
+      | otherwise = error $ "test expected to begin with preamble: " ++ preamble
+
+    before :: Int -> [a] -> [a]
+    before = take
+
+    after :: Int -> [a] -> [a]
+    after idx = drop (idx + 1)
+
+-- | Create a test case from a .wasp/.golden pair, running runCompletionTest
+-- on the .wasp file to get the .golden output. See 'readCompletionTest' for
+-- the format of a test case.
+makeCompletionTest :: FilePath -> TestTree
```

Comment:
Ok, so this is given an test input file and based on it creates a golden test for completion. It will create golden file with same name but with .golden extension next to the test input file. Makes sense to me!

I would change name to `makeGoldenCompletionTest` or `makeGoldenCompletionTestFromTestInputFile`. Maybe the last one is the best even.

I would also consider using StrongPath -> it has to be File, it has to be System, and now is it absolute or relative -> I guess we don't care? But, even if we don't care, would be nice if user of this method knows we don't care, instead of wondering.
So, for example we can do `Path' a File'` then instead of `FilePath`.

As for comments I would also update them a tiny bit:
1. "create a test case from a .wasp/.golden pair" -> That confused me a bit. .wasp/.golden looked like a path to me. So we could say "from a pair of <test_name>.wasp, <test_name>.golden files". Although, we could argue it really makes a test from just the input file. But it does also use golden file if it is there, true. Ok. Also, since this function uses runCompletionTest, I would put it first -> first I learn about how to make a completion test, then about how it actually runs. For the sake of the "story".

Reply:
> I would change name to makeGoldenCompletionTest or makeGoldenCompletionTestFromTestInputFile. Maybe the last one is the best even.

Changed name to `makeGoldenCompletionTest`.

> As for comments I would also update them a tiny bit:

👍 

> I would also consider using StrongPath

I just tried that out and it doesn't feel worth it to me. It adds a lot of noise where the paths are used/created from and only improves the readability of `makeGoldenCompletionTest`. I've added to the doc comment for `makeGoldenCompletionTest` what the `FilePath` is expected to be, let me know if you think that is good enough.

Reply:
StrongPath -> ok sure, if you feel it clutters the whole thing too much, then comments can also do the trick! It is localized function anyway, relatively isolated.